util/src/openssl_utils.cpp

```diff
@@ -0,0 +1,258 @@
+                                         uint32_t& remote_peer_id,
+                                         string& conn_type,
+                                         bool use_unified_certs) {
+  // First get the source ID
+  static constexpr size_t SIZE = 512;
+  string subject(SIZE, 0);
+  X509_NAME_oneline(X509_get_subject_name(cert_to_verify), subject.data(), SIZE);
+
+  int peerIdPrefixLength = 3;
+  std::regex r("OU=\\d*", std::regex_constants::icase);
+  std::smatch sm;
+  regex_search(subject, sm, r);
+  if (sm.length() <= peerIdPrefixLength) {
+    LOG_ERROR(OPENSSL_LOG, "OU not found or empty: " << subject);
+    return false;
+  }
+
+  auto remPeer = sm.str().substr(peerIdPrefixLength, sm.str().length() - peerIdPrefixLength);
+  if (0 == remPeer.length()) {
+    LOG_ERROR(OPENSSL_LOG, "OU empty " << subject);
+    return false;
+  }
+
+  uint32_t remotePeerId;
+  try {
+    remotePeerId = stoul(remPeer, nullptr);
+  } catch (const std::invalid_argument& ia) {
+    LOG_ERROR(OPENSSL_LOG, "cannot convert OU, " << subject << ", " << ia.what());
+    return false;
+  } catch (const std::out_of_range& e) {
+    LOG_ERROR(OPENSSL_LOG, "cannot convert OU, " << subject << ", " << e.what());
+    return false;
+  }
+  remote_peer_id = remotePeerId;
+  string CN;
+  CN.resize(SIZE);
+  X509_NAME_get_text_by_NID(X509_get_subject_name(cert_to_verify), NID_commonName, CN.data(), SIZE);
+  string cert_type = "server";
+  if (CN.find("cli") != string::npos) cert_type = "client";
+  conn_type = cert_type;
+
+  // Get the local stored certificate for this peer
+  string local_cert_path =
+      (use_unified_certs)
+          ? cert_root_directory + "/" + std::to_string(remotePeerId) + "/" + "node.cert"
+          : cert_root_directory + "/" + std::to_string(remotePeerId) + "/" + cert_type + "/" + cert_type + ".cert";
+  auto deleter = [](FILE* fp) {
+    if (fp) fclose(fp);
+  };
+  std::unique_ptr<FILE, decltype(deleter)> fp(fopen(local_cert_path.c_str(), "r"), deleter);
+  if (!fp) {
+    LOG_ERROR(OPENSSL_LOG, "Certificate file not found, path: " << local_cert_path);
+    return false;
+  }
+
+  UniqueOpenSSLX509 localCert(PEM_read_X509(fp.get(), NULL, NULL, NULL));
+  if (!localCert) {
+    LOG_ERROR(OPENSSL_LOG, "Cannot parse certificate, path: " << local_cert_path);
+    return false;
+  }
+
+  // this is actual comparison, compares hash of 2 certs
+  bool res = (X509_cmp(cert_to_verify, localCert.get()) == 0);
+  return res;
+}
+
+pair<string, string> OpenSSLCryptoImpl::generateEdDSAKeyPair(const KeyFormat fmt) const {
+  UniquePKEY edPkey;
+  UniqueOpenSSLPKEYContext edPkeyCtx(EVP_PKEY_CTX_new_id(NID_ED25519, nullptr));
+
+  ConcordAssertNE(edPkeyCtx, nullptr);
+
+  ConcordAssertEQ(1, EVP_PKEY_keygen_init(edPkeyCtx.get()));
+  ConcordAssertEQ(
+      1, EVP_PKEY_keygen(edPkeyCtx.get(), reinterpret_cast<EVP_PKEY**>(&edPkey)));  // Generate EdDSA key 'edPkey'.
+
+  unsigned char privKey[EdDSASignatureByteSize]{};
```

Comment:
is this code duplicated?
I remember commenting on similar looking code to use std::array.

Reply:
The current code is fine for this implementation because we need `privlen` and `publen` variables in `EVP_PKEY_get_raw_private_key()` & `EVP_PKEY_get_raw_public_key()` as the last param.
However, `concordUtils::bufferToHex()` is changed to `boost::algorithm::hex()`.

Reply:
so use an additional variable, but don't use c-style arrays.
And again, I know for a fact that there are other places in the code which generate eddsa key pairs, please deduplicate the code.

Reply:
you can use local variables and std::array

Reply:
`privKey`, `pubKey`, `privlen` & `publen` are local variables.
I suggest to keep `unsigned char privKey[]` as is and not convert to `std::array<uint8_t>`. This will save us from converting to `const char *` in two places. We are already doing in one place i.e. in `boost::algorithm::hex()`.
wdyt?

Reply:
Resolved in https://github.com/arc-vmware/concord-bft/pull/6/commits/974147e0cb0d5ec81db4edb8fdb3b12336135cea